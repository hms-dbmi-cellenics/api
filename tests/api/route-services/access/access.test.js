const AWSMock = require('aws-sdk-mock');
const fake = require('../../../test-utils/constants');

const {
  mockDocClientQuery,
  mockDocClientBatchWrite,
  mockDynamoScan,
} = require('../../../test-utils/mockAWSServices');
const AccessService = require('../../../../src/api/route-services/access');

jest.mock('../../../../src/api/route-services/experiment');
jest.mock('../../../../src/utils/authMiddlewares');

describe('tests for the projects service', () => {
  afterEach(() => {
    AWSMock.restore('DynamoDB');
  });

  test('getExperimentEntries calls doc client', async () => {
    const items = [
      {
        userId: fake.USER.sub,
        experimentId: fake.EXPERIMENT_ID,
        projectId: fake.PROJECT_ID,
        role: 'OWNER',
      },
    ];
    const deleteSpy = mockDocClientQuery(items);


    const as = new AccessService();
    const responseItems = await as.getExperimentEntries('table', fake.EXPERIMENT_ID);

    expect(deleteSpy).toHaveBeenCalledTimes(1);
    expect(deleteSpy.mock.calls).toMatchSnapshot();
    expect(items).toEqual(responseItems);
  });

  test('deleteExperimentEntries calls doc client', async () => {
    const items = [
      {
        userId: fake.USER.sub,
        experimentId: fake.EXPERIMENT_ID,
        projectId: fake.PROJECT_ID,
        role: 'OWNER',
      },
    ];
    const deleteSpy = mockDocClientBatchWrite();

    const as = new AccessService();

    await as.deleteExperimentEntries(as.userAccessTableName, fake.EXPERIMENT_ID, items);

    expect(deleteSpy).toHaveBeenCalledTimes(1);
    expect(deleteSpy.mock.calls).toMatchSnapshot();

    deleteSpy.mockClear();

    await as.deleteExperimentEntries(as.inviteAccessTableName, fake.EXPERIMENT_ID, items);

    expect(deleteSpy).toHaveBeenCalledTimes(1);
    expect(deleteSpy.mock.calls).toMatchSnapshot();
  });

  test('GetAccessibleProjects gets all the projects across many pages of scan results', async () => {
    const projectIds1 = [
      {
        projectId: 'project-1',
      },
      {
        projectId: 'project-2',
      },
      {
        projectId: 'project-3',
      },
    ];

    const projectIds2 = [
      {
        projectId: 'project-4',
      },
      {
        projectId: 'project-5',
      },
      {
        projectId: 'project-6',
      },
    ];

    mockDynamoScan([projectIds1, projectIds2]);

    const expectedResult = [...projectIds1, ...projectIds2].map(({ projectId }) => projectId);

    const accessService = new AccessService();
    const user = { sub: 'mockSubject' };

    const res = await accessService.getAccessibleProjects(user);

    expect(res).toEqual(expectedResult);
  });

  test('GetProjects gets all the projects across many pages of scan results when the first page is empty', async () => {
    const emptyProjectIds = [];

    const projectIds2 = [
      {
        projectId: 'project-3',
      },
      {
        projectId: 'project-4',
      },
      {
        projectId: 'project-5',
      },
      {
        projectId: 'project-6',
      },
    ];

    mockDynamoScan([emptyProjectIds, projectIds2]);

    const expectedResult = [...emptyProjectIds, ...projectIds2].map(({ projectId }) => projectId);

    const accessService = new AccessService();
    const user = { sub: 'mockSubject' };

    const res = await accessService.getAccessibleProjects(user);

    expect(res).toEqual(expectedResult);
  });

  test('GetProjects removes duplicate projectIds', async () => {
    const projectIds1 = [
      {
        projectId: 'project-1',
      },
      {
        projectId: 'project-2',
      },
      {
        projectId: 'project-3',
      },
      {
        projectId: 'project-4',
      },
      {
        projectId: 'project-1',
      },
    ];

    mockDynamoScan([projectIds1]);

    const allProjectIds = projectIds1.map(({ projectId }) => projectId);
    const expectedResult = [...new Set(allProjectIds)];

    const accessService = new AccessService();
    const user = { sub: 'mockSubject' };

    const res = await accessService.getAccessibleProjects(user);

    expect(res).toEqual(expectedResult);
  });

  test('deleteExperiment calls get and delete', async () => {
    const items = [
      {
        userId: fake.USER.sub,
        experimentId: fake.EXPERIMENT_ID,
        projectId: fake.PROJECT_ID,
        role: 'OWNER',
      },
    ];

    const getMock = jest.fn(() => items);
    const deleteMock = jest.fn();


    const as = new AccessService();
    as.getExperimentEntries = getMock;
    as.deleteExperimentEntries = deleteMock;
    await as.deleteExperiment(fake.EXPERIMENT_ID);

    expect(getMock).toHaveBeenCalledTimes(2);
    expect(deleteMock).toHaveBeenCalledTimes(2);
  });

  test('deleteExperiment does not call delete if there are no items', async () => {
    const getMock = jest.fn(() => []);
    const deleteMock = jest.fn();


    const as = new AccessService();
    as.getExperimentEntries = getMock;
    as.deleteExperimentEntries = deleteMock;
    await as.deleteExperiment(fake.EXPERIMENT_ID);

    expect(getMock).toHaveBeenCalledTimes(2);
    expect(deleteMock).toHaveBeenCalledTimes(0);
  });
});
