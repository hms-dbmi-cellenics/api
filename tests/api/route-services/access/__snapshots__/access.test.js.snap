// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`tests for the projects service deleteExperimentEntries calls doc client 1`] = `
Array [
  Array [
    Object {
      "RequestItems": Object {
        "user-access-test": Array [
          Object {
            "DeleteRequest": Object {
              "Key": Object {
                "experimentId": "experimentid11111111111111111111",
                "userId": "032cdb44-0cd3-4d58-af21-850kp0b95ac7",
              },
            },
          },
        ],
      },
    },
  ],
]
`;

exports[`tests for the projects service deleteExperimentEntries calls doc client 2`] = `
Array [
  Array [
    Object {
      "RequestItems": Object {
        "invite-access-test": Array [
          Object {
            "DeleteRequest": Object {
              "Key": Object {
                "experimentId": "experimentid11111111111111111111",
                "userEmail": undefined,
              },
            },
          },
        ],
      },
    },
  ],
]
`;

exports[`tests for the projects service getExperimentEntries calls doc client 1`] = `
Array [
  Array [
    Object {
      "ExpressionAttributeValues": Object {
        ":hkey": "experimentid11111111111111111111",
      },
      "KeyConditionExpression": "experimentId = :hkey",
      "TableName": "table",
    },
  ],
]
`;
